pkg/operator/missingstaticpodcontroller/missingstaticpodcontroller.go

```diff
@@ -0,0 +1,209 @@
+			kubeInformersForTargetNamespace.Core().V1().Pods().Informer(),
+		).
+		WithSync(c.sync).
+		WithSyncDegradedOnError(operatorClient).
+		ToController("MissingStaticPodController", eventRecorder)
+}
+
+func (c MissingStaticPodController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	_, originalOperatorStatus, _, err := c.operatorClient.GetStaticPodOperatorState()
+	if err != nil {
+		return err
+	}
+
+	installerPods, err := c.podLister.List(labels.SelectorFromSet(labels.Set{"app": "installer"}))
+	if err != nil {
+		return err
+	}
+
+	// get the most recent installer pod for each node
+	latestInstallerPodsByNode := getMostRecentInstallerPodByNode(installerPods)
+	for node, latestInstallerPodOnNode := range latestInstallerPodsByNode {
+		installerPodRevision, err := getInstallerPodRevision(latestInstallerPodOnNode)
+		if err != nil {
+			// we expect every installer pod to have a installerPodRevision in its name, unexpected error here
+			return fmt.Errorf("failed to get installerPodRevision for installer pod %q - %w", latestInstallerPodOnNode.Name, err)
+		}
+
+		finishedAt, ok := installerPodFinishedAt(latestInstallerPodOnNode)
+		if !ok {
+			// either it's in the process of running, or it ran into an error
+			continue
+		}
+		threshold := terminationGracePeriodSeconds(latestInstallerPodOnNode) + 10*time.Second
+		staticPodRevisionOnThisNode := getStaticPodCurrentRevision(node, originalOperatorStatus)
+		if time.Since(finishedAt) > threshold &&
+			staticPodRevisionOnThisNode < installerPodRevision {
+			// if we are here:
+			//  a: the latest installer pod successfully completed at finishedAt
+			//  b: it has been more than 'terminationGracePeriodSeconds' + 10s since the
+			//     installer pod has completed at finishedAt
+			//  c. the static pod is not at correct installerPodRevision yet
+			// then all of the above conditions are true, so we should go degraded
+			//
+			// TODO: set a degraded condition
+			syncCtx.Recorder().Eventf("MissingStaticPod", "static pod lifecycle failure - installer pod: %q with revision: %d completed at: %s, static pod revision: %d",
+				latestInstallerPodOnNode.Name, installerPodRevision, finishedAt, staticPodRevisionOnThisNode)
+		}
+	}
+
+	return nil
+}
+
+func getStaticPodCurrentRevision(node string, status *operatorv1.StaticPodOperatorStatus) int {
+	var nodeStatus *operatorv1.NodeStatus
+	for i := range status.NodeStatuses {
+		if status.NodeStatuses[i].NodeName == node {
+			nodeStatus = &status.NodeStatuses[i]
+			break
+		}
+	}
+
+	if nodeStatus == nil {
+		return 0
+	}
+	return int(nodeStatus.CurrentRevision)
+}
+
+func terminationGracePeriodSeconds(pod *corev1.Pod) time.Duration {
+	value := pod.Spec.TerminationGracePeriodSeconds
+	if value == nil {
+		return 0
+	}
+
+	return time.Duration(*value * int64(time.Second))
+}
+
+func getMostRecentInstallerPodByNode(pods []*corev1.Pod) map[string]*corev1.Pod {
+	mostRecentInstallerPodByNode := map[string]*corev1.Pod{}
```

Comment:
this will not work if there are multiple installer pods inside of the same namespace. Are we certain that this will never be the case?

Reply:
we discussed it offline, Damien wanted to know if an installer-pod can install more than one pod.